### PR TITLE
imp: require stricter usage of app version in packet

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

closes: #7122

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/04-channel/types/packet_test.go [hunk 1]
```diff
@@ -57,6 +57,9 @@ func TestPacketValidateBasic(t *testing.T) {
 		{types.NewPacket(validPacketData, 1, portid, chanid, cpportid, cpchanid, disabledTimeout, 0), false, "disabled both timeout height and timestamp"},
 		{types.NewPacket(validPacketData, 1, portid, chanid, cpportid, cpchanid, disabledTimeout, timeoutTimestamp), true, "disabled timeout height, valid timeout timestamp"},
 		{types.NewPacket(validPacketData, 1, portid, chanid, cpportid, cpchanid, timeoutHeight, 0), true, "disabled timeout timestamp, valid timeout height"},
+		{types.NewPacketWithVersion(validPacketData, 1, portid, chanid, cpportid, cpchanid, timeoutHeight, timeoutTimestamp, "version"), true, "valid v2 packet"},
```
Comment (colin-axner) [inline modules/core/04-channel/types/packet_test.go:60]: feel like this tests could be cleaned up on main, having the test name last is confusing + could use expected errors

File: modules/core/04-channel/types/packet_test.go [hunk 2]
```diff
@@ -57,6 +57,9 @@ func TestPacketValidateBasic(t *testing.T) {
...
+		{types.Packet{1, portid, chanid, cpportid, cpchanid, validPacketData, timeoutHeight, timeoutTimestamp, types.IBC_VERSION_1, "version"}, false, "invalid specifying of app version with unspecified protocol version"},
```
Comment (colin-axner) [inline modules/core/04-channel/types/packet_test.go:61]:
```suggestion
		{types.Packet{1, portid, chanid, cpportid, cpchanid, validPacketData, timeoutHeight, timeoutTimestamp, types.IBC_VERSION_1, "version"}, false, "invalid specifying of app version with protocol version 1"},
```

File: modules/core/04-channel/types/packet.go [hunk 3]
```diff
@@ -195,6 +196,9 @@ func (p Packet) ValidateBasic() error {
 	if len(p.Data) == 0 {
 		return errorsmod.Wrap(ErrInvalidPacket, "packet data bytes cannot be empty")
 	}
+	if p.AppVersion != "" && slices.Contains([]IBCVersion{IBC_VERSION_UNSPECIFIED, IBC_VERSION_1}, p.ProtocolVersion) {
+		return errorsmod.Wrap(ErrInvalidPacket, "app version cannot be specified when packet uses protocol version 1")
```
Comment (crodriguezvega) [inline modules/core/04-channel/types/packet.go:200]: Just to cover the case in the error message where protocol version is empty.

```suggestion
		return errorsmod.Wrapf(ErrInvalidPacket, "app version cannot be specified when packet does not use protocol %s", IBC_VERSION_2)
```

**Review by crodriguezvega — APPROVED** [review]

**Comment by crodriguezvega** [comment]
Since we are merging to feature branch, I will go ahead and merge with one approval.

**Pull request merged**